src/core/qgsvectorlayer.cpp

```diff
@@ -2243,10 +2248,34 @@ bool QgsVectorLayer::changeGeometry( QgsFeatureId fid, const QgsGeometry &geom )
 
 bool QgsVectorLayer::changeAttributeValue( QgsFeatureId fid, int field, const QVariant &newValue, const QVariant &oldValue )
 {
-  if ( !mEditBuffer || !mDataProvider )
-    return false;
+  if ( fields().fieldOrigin( field ) == QgsFields::OriginJoin )
+  {
+    int srcFieldIndex;
+    const QgsVectorLayerJoinInfo *info = mJoinBuffer->joinForFieldIndex( field, fields(), srcFieldIndex );
+    if ( info && info->joinLayer() && info->isEditable() )
+    {
+      QgsFeature feature = getFeature( fid );
+
+      if ( !feature.isValid() )
+        return false;
+
+      QgsFeature joinFeature = mJoinBuffer->joinedFeatureOf( info, feature );
 
-  return mEditBuffer->changeAttributeValue( fid, field, newValue, oldValue );
+      if ( joinFeature.isValid() )
+        return info->joinLayer()->changeAttributeValue( joinFeature.id(), srcFieldIndex, newValue, oldValue );
```

Comment:
> but in this case we need to give helpful feedback explaining why the edit failed. Silently failing isn't going to do us any favours with our users ;)

Indeed. I added an icon on the right of editor widgets in attribute form. Below an example where a feature is being created in the target layer with :
- **p1** not an editable join
- **p2** an editable join and *Upsert on edit* is ativated but the layer is not currently editable
- **p4** an editable join but *Upsert on edit* is not activated

@nyalldawson @haubourg Let me know of it suits you
![6](https://user-images.githubusercontent.com/9266424/28611276-24c5cc9c-71e3-11e7-8d88-d9c7857abcfa.png)